sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/ThroughputRequestThrottler.java

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.throughputControl;
+
+import com.azure.cosmos.BridgeInternal;
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.implementation.HttpConstants;
+import com.azure.cosmos.implementation.RequestRateTooLargeException;
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.RxDocumentServiceResponse;
+import com.azure.cosmos.implementation.Utils;
+import com.azure.cosmos.implementation.directconnectivity.StoreResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This is the place where we tracking the RU usage, and make decision whether we should block the request.
+ */
+public class ThroughputRequestThrottler {
+    private static final Logger logger = LoggerFactory.getLogger(ThroughputRequestThrottler.class);
+
+    private final AtomicReference<Double> availableThroughput;
+    private final AtomicReference<Double> scheduledThroughput;
+
+    public ThroughputRequestThrottler(double scheduledThroughput) {
+        this.availableThroughput = new AtomicReference<>(scheduledThroughput);
+        this.scheduledThroughput = new AtomicReference<>(scheduledThroughput);
+    }
+
+    public Mono<Void> renewThroughputUsageCycle(double scheduledThroughput) {
+        this.scheduledThroughput.set(scheduledThroughput);
+        this.updateAvailableThroughput();
+        return Mono.empty();
+    }
+
+    private void updateAvailableThroughput() {
+        // The base rule is: If RU is overused during the current cycle, the over used part will be deducted from the next cyclle
+        // If RU is not fully utilized during the current cycle, it will be void.
+        // So the available throughput for each cycle will be less than scheduled throughput.
+        this.availableThroughput.getAndAccumulate(this.scheduledThroughput.get(), (available, refill) -> Math.min(available,0) + refill);
+    }
+
+    public <T> Mono<T> processRequest(RxDocumentServiceRequest request, Mono<T> originalRequestMono) {
+        if (this.availableThroughput.get() > 0) {
+            return originalRequestMono
+                .doOnSuccess(response -> {
+                    this.trackRequestCharge(response);
+                })
+                .onErrorResume(throwable -> {
```

Comment:
If we are going to throw error on `onErrorResume`, we can just use `doOnError`. 

`onErrorResume()` is used in cases where you want to fallback to a different publisher in case of errors, but here you just want to achieve a side effect. 
From documentation of `onErrorResume()` -> 


/**
	 * Subscribe to a fallback publisher when any error occurs, using a function to
	 * choose the fallback depending on the error.
	 *
**/

Reply:
Using onErrorResume because somehow previously when I do tests, doOnError is not implemented when exception happens. 

But retested and doOnError works, so has updated back to doOnError